_posts/2023-01-14-trivially-relocatable.md

```diff
@@ -0,0 +1,656 @@
+macro.
+
+```cpp
+struct sus_if_clang([[clang::trivial_abi]]) S {
+    Thing<int> thing;
+    int i;
+    sus_class_assert_trivial_relocatable_types(
+        unsafe_fn,
+        decltype(thing),
+        decltype(i));
+};
+```
+
+#### sus_class_trivial_relocatable(unsafe_fn)
+
+The simplest but most risky macro is `sus_class_trivial_relocatable()`. This macro is like
+`sus_class_assert_trivial_relocatable_types()` but without the additional help of the assertion
+against the member types. When using this macro, the type should also be annotated with the
+`[[clang::trivial_abi]]` attribute.
+
+```cpp
+struct sus_if_clang([[clang::trivial_abi]]) S {
+    Thing<int> thing;
+    int i;
+    sus_class_trivial_relocatable(unsafe_fn);
+};
+```
+
+#### sus_class_maybe_trivial_relocatable_types(unsafe_fn, types...)
+
+The format of the macro is just like `sus_class_assert_trivial_relocatable_types()` but if any
+type given to the macro is not trivially relocatable, the containing type will also not be.
+
+Specifically, this allows a type to opt into being trivially relocatable if all of its members are
+trivially relocatable, including template parameter types.
+
+This macro is probably only worth using in a template, as otherwise the types are either known to
+be trivially relocatable or to not, and the `sus_class_assert_trivial_relocatable_types()` macro
+could be used in the former case. And since the condition can evaluate to false, the use of
+`[[clang::trivial_abi]]` on such a class type would be a bug.
+
+```cpp
+template <class T>
+struct S {
+    Thing<T> thing;
+    T t;
+    sus_class_maybe_trivial_relocatable_types(
+        unsafe_fn,
+        decltype(thing),
+        decltype(t));
+};
+```
+
+The behaviour of `sus_class_maybe_trivial_relocatable_types()` is much like the extensions to
+the compiler proposed in
+[P1144R6](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1144r6.html). 
+
+#### sus_class_trivial_relocatable_value(unsafe_fn, bool)
+
+The `sus_class_trivial_relocatable_value()` macro receives a boolean argument that will be
+constant evaluated and used to determine if the type is ultimately marked as trivially relocatable
+or not. This is useful when the condition is more complex than just whether the members of the type
+are themselves trivially relocatable, but the caller can make use of
+`sus::mem::relocate_by_memcpy<T>` to check members as well.
+
+Since the condition can evaluate to false, the use of `[[clang::trivial_abi]]` on such a class type
+would be a bug.
+
+```cpp
+template <class T>
+struct S {
+    Thing<T> thing;
+    T t;
+    sus_class_trivial_relocatable_value(
+        unsafe_fn,
+        StuffAbout<T> &&
+        sus::mem::relocate_by_memcpy<decltype(thing)> &&
+        sus::mem::relocate_by_memcpy<decltype(t)>);
+};
+```
+
+The `sus_class_trivial_relocatable_value()` macro is most similar to the proposed

```

Comment:
Two thoughts about naming:
- I am biased towards trivialLY relocatable rather than just "trivial"
- Given the three other macros, I feel like sus_class_trivial_relocatable_if() would be a little more immediately understandable to someone not as familiar.

Reply:
Internal debate about renaming things even more mid blog post or not ensues... I like both suggestions.

Reply:
Ok incoming renames heh. Stupid git permalinks..